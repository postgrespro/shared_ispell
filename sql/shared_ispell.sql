CREATE EXTENSION shared_ispell;

-- Test ISpell dictionary with ispell affix file
CREATE TEXT SEARCH DICTIONARY shared_ispell (
                        Template=shared_ispell,
                        DictFile=ispell_sample,
                        AffFile=ispell_sample,
                        Stopwords=english
);

SELECT ts_lexize('shared_ispell', 'skies');
SELECT ts_lexize('shared_ispell', 'bookings');
SELECT ts_lexize('shared_ispell', 'booking');
SELECT ts_lexize('shared_ispell', 'foot');
SELECT ts_lexize('shared_ispell', 'foots');
SELECT ts_lexize('shared_ispell', 'rebookings');
SELECT ts_lexize('shared_ispell', 'rebooking');
SELECT ts_lexize('shared_ispell', 'unbookings');
SELECT ts_lexize('shared_ispell', 'unbooking');
SELECT ts_lexize('shared_ispell', 'unbook');

SELECT ts_lexize('shared_ispell', 'footklubber');
SELECT ts_lexize('shared_ispell', 'footballklubber');
SELECT ts_lexize('shared_ispell', 'ballyklubber');
SELECT ts_lexize('shared_ispell', 'footballyklubber');

-- Test ISpell dictionary with hunspell affix file
CREATE TEXT SEARCH DICTIONARY shared_hunspell (
                        Template=shared_ispell,
                        DictFile=ispell_sample,
                        AffFile=hunspell_sample
);

SELECT ts_lexize('shared_hunspell', 'skies');
SELECT ts_lexize('shared_hunspell', 'bookings');
SELECT ts_lexize('shared_hunspell', 'booking');
SELECT ts_lexize('shared_hunspell', 'foot');
SELECT ts_lexize('shared_hunspell', 'foots');
SELECT ts_lexize('shared_hunspell', 'rebookings');
SELECT ts_lexize('shared_hunspell', 'rebooking');
SELECT ts_lexize('shared_hunspell', 'unbookings');
SELECT ts_lexize('shared_hunspell', 'unbooking');
SELECT ts_lexize('shared_hunspell', 'unbook');

SELECT ts_lexize('shared_hunspell', 'footklubber');
SELECT ts_lexize('shared_hunspell', 'footballklubber');
SELECT ts_lexize('shared_hunspell', 'ballyklubber');
SELECT ts_lexize('shared_hunspell', 'footballyklubber');

SELECT dict_name, affix_name, words, affixes FROM shared_ispell_dicts();
SELECT stop_name, words FROM shared_ispell_stoplists();

SELECT shared_ispell_reset();

SELECT ts_lexize('shared_ispell', 'skies');
SELECT ts_lexize('shared_hunspell', 'skies');
SELECT ts_lexize('shared_hunspell', 'skies');
