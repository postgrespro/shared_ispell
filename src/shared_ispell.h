#ifndef __SHARED_ISPELL_H__
#define __SHARED_ISPELL_H__

#include "storage/lwlock.h"
#include "utils/memutils.h"
#include "utils/timestamp.h"
#include "tsearch/dicts/spell.h"
#include "tsearch/ts_public.h"

/* This segment is initialized in the first process that accesses it (see
 * ispell_shmem_startup function).
 */
#define SEGMENT_NAME "shared_ispell"

#define MAXLEN 255

typedef struct SharedIspellDict
{
	/* this is used for selecting the dictionary */
	char   *dictFile;
	char   *affixFile;
	int		nbytes;
	int		nwords;

	/* next dictionary in the chain (essentially a linked list) */
	struct SharedIspellDict *next;

	IspellDict dict;
} SharedIspellDict;

typedef struct SharedStopList
{
	char   *stopFile;
	int		nbytes;

	struct SharedStopList *next;

	StopList stop;
} SharedStopList;

/* used to allocate memory in the shared segment */
typedef struct SegmentInfo
{
	LWLockId	lock;
	char	   *firstfree;			/* first free address (always maxaligned) */
	size_t		available;			/* free space remaining at firstfree */
	instr_time	lastReset;			/* last reset of the dictionary */

	/* the shared segment (info and data) */
	SharedIspellDict *shdict;
	SharedStopList	 *shstop;
} SegmentInfo;

/* used to keep track of dictionary in each backend */
typedef struct DictInfo
{
	instr_time	lookup;

	char dictFile[MAXLEN];
	char affixFile[MAXLEN];
	char stopFile[MAXLEN];

	/* We split word list and affix list.
	 * In shdict we store a word list, word list will be stored in shared segment.
	 * In dict we store an affix list in each process.
	 */
	SharedIspellDict   *shdict;
	IspellDict			dict;
	SharedStopList	   *shstop;

	/* MemoryContext of dict local content */
	MemoryContext		infoCntx;
} DictInfo;

#endif
