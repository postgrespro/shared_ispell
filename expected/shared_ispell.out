CREATE EXTENSION shared_ispell;
-- Test ISpell dictionary with ispell affix file
CREATE TEXT SEARCH DICTIONARY shared_ispell (
                        Template=shared_ispell,
                        DictFile=ispell_sample,
                        AffFile=ispell_sample,
                        Stopwords=english
);
SELECT ts_lexize('shared_ispell', 'skies');
 ts_lexize 
-----------
 {sky}
(1 row)

SELECT ts_lexize('shared_ispell', 'bookings');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_ispell', 'booking');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_ispell', 'foot');
 ts_lexize 
-----------
 {foot}
(1 row)

SELECT ts_lexize('shared_ispell', 'foots');
 ts_lexize 
-----------
 {foot}
(1 row)

SELECT ts_lexize('shared_ispell', 'rebookings');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_ispell', 'rebooking');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_ispell', 'unbookings');
 ts_lexize 
-----------
 {book}
(1 row)

SELECT ts_lexize('shared_ispell', 'unbooking');
 ts_lexize 
-----------
 {book}
(1 row)

SELECT ts_lexize('shared_ispell', 'unbook');
 ts_lexize 
-----------
 {book}
(1 row)

SELECT ts_lexize('shared_ispell', 'footklubber');
   ts_lexize    
----------------
 {foot,klubber}
(1 row)

SELECT ts_lexize('shared_ispell', 'footballklubber');
                      ts_lexize                       
------------------------------------------------------
 {footballklubber,foot,ball,klubber,football,klubber}
(1 row)

SELECT ts_lexize('shared_ispell', 'ballyklubber');
   ts_lexize    
----------------
 {ball,klubber}
(1 row)

SELECT ts_lexize('shared_ispell', 'footballyklubber');
      ts_lexize      
---------------------
 {foot,ball,klubber}
(1 row)

-- Test ISpell dictionary with hunspell affix file
CREATE TEXT SEARCH DICTIONARY shared_hunspell (
                        Template=shared_ispell,
                        DictFile=ispell_sample,
                        AffFile=hunspell_sample
);
SELECT ts_lexize('shared_hunspell', 'skies');
 ts_lexize 
-----------
 {sky}
(1 row)

SELECT ts_lexize('shared_hunspell', 'bookings');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_hunspell', 'booking');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_hunspell', 'foot');
 ts_lexize 
-----------
 {foot}
(1 row)

SELECT ts_lexize('shared_hunspell', 'foots');
 ts_lexize 
-----------
 {foot}
(1 row)

SELECT ts_lexize('shared_hunspell', 'rebookings');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_hunspell', 'rebooking');
   ts_lexize    
----------------
 {booking,book}
(1 row)

SELECT ts_lexize('shared_hunspell', 'unbookings');
 ts_lexize 
-----------
 {book}
(1 row)

SELECT ts_lexize('shared_hunspell', 'unbooking');
 ts_lexize 
-----------
 {book}
(1 row)

SELECT ts_lexize('shared_hunspell', 'unbook');
 ts_lexize 
-----------
 {book}
(1 row)

SELECT ts_lexize('shared_hunspell', 'footklubber');
   ts_lexize    
----------------
 {foot,klubber}
(1 row)

SELECT ts_lexize('shared_hunspell', 'footballklubber');
                      ts_lexize                       
------------------------------------------------------
 {footballklubber,foot,ball,klubber,football,klubber}
(1 row)

SELECT ts_lexize('shared_hunspell', 'ballyklubber');
   ts_lexize    
----------------
 {ball,klubber}
(1 row)

SELECT ts_lexize('shared_hunspell', 'footballyklubber');
      ts_lexize      
---------------------
 {foot,ball,klubber}
(1 row)

SELECT dict_name, affix_name, words, affixes FROM shared_ispell_dicts();
   dict_name   |   affix_name    | words | affixes 
---------------+-----------------+-------+---------
 ispell_sample | hunspell_sample |     8 |       7
 ispell_sample | ispell_sample   |     8 |       7
(2 rows)

SELECT stop_name, words FROM shared_ispell_stoplists();
 stop_name | words 
-----------+-------
 english   |   127
(1 row)

SELECT shared_ispell_reset();
 shared_ispell_reset 
---------------------
 
(1 row)

SELECT ts_lexize('shared_ispell', 'skies');
 ts_lexize 
-----------
 {sky}
(1 row)

SELECT ts_lexize('shared_hunspell', 'skies');
 ts_lexize 
-----------
 {sky}
(1 row)

SELECT ts_lexize('shared_hunspell', 'skies');
 ts_lexize 
-----------
 {sky}
(1 row)

